dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/stack/transport/sip/netty/SipUdpConnLink.java

```diff
@@ -0,0 +1,492 @@
+		}
+		if (message == null) {
+			throw new IOException("message is null in SipUdpConnLink.sendNow");
+		}
+		if (m_outboundBuffer != null) {
+			throw new IOException("previous send not completed in SipUdpConnLink.sendNow");
+		}
+
+		SipUdpConnection sipConnection = (SipUdpConnection) messageContext.getSipConnection();
+		String remoteHost = sipConnection.getRemoteHost();
+		int remotePort = sipConnection.getRemotePort();
+		InetSocketAddress address = InetAddressCache.getInetSocketAddress(remoteHost, remotePort);
+
+		m_outboundBuffer = BaseConnection.stackBufferToByteBuf(message);
+		
+		DatagramPacket pkt = new DatagramPacket(m_outboundBuffer, address);
+		final ChannelFuture writeFuture = sipConnection.getChannel().writeAndFlush(pkt);
+		boolean complete = writeFuture.isDone();
+		
+		writeFuture.addListener(this);
+		
+		return complete;
+	}
+
+	// ------------------------------
+	// SIPReadCallback implementation
+	// ------------------------------
+
+	/**
+	 * called by the channel framework when a new message arrives
+	 */
+	public void complete(SipMessageByteBuffer buffer, InetSocketAddress senderAddr) {
+		boolean drop = Dispatcher.instance().isOverLoaded();
+		if (drop) {
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(this, tc, "complete", "Warning: dropping request under overloaded situation");
+			}
+		}
+		if (!m_connected) {
+			m_connected = true;
+		}
+		if (senderAddr == null) {
+			if (c_logger.isInfoEnabled()) {
+				c_logger.info("complete", null, "A message in order to use UDP connection link");
+			}
+		} else  {
+			BaseConnection bc = new SipUdpConnection(m_channel, this, m_channel.getChannel());
+			bc.setRemoteAddress(senderAddr);
+			bc.messageReceived(buffer);
+		} 
+		
+	}
+
+	public void error(Throwable e) {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(this, tc, "SipUdpConnLink(read).error", e);
+		}
+		close(e);
+	}
+
+	// -------------------------------
+	// SIPWriteCallback implementation
+	// -------------------------------
+
+	/**
+	 * called by the channel framework when previous send completed
+	 * 
+	 */
+	public void complete() {
+		m_sendThread.sendComplete();
+	}
+
+	/**
+	 * called when send completed to release the outbound buffer, either from the
+	 * send thread (immediate) or from the channel thread (asynchronous)
+	 */
+	void releaseOutboundBuffer() {
+		// TODO do we need to release a buffer ?
+		// m_outboundBuffer.release();
+		m_outboundBuffer = null;
+	}
+
+	public void close(Throwable e) {
+		if (!m_connected) {
+			// avoid double-close
+			return;
+		}
+		m_connected = false;
+
+		// remove this conn link from the global table
+		s_instances.remove(m_channel.getListeningPoint());
+
+		// tell the channel that we're closed. this will cause the next message
+		// to create a new instance of this class.
+		m_channel.connectionClosed();
```

Comment:
I am worried about Netty required cleanup as we have:
```
	void connectionClosed() {
		m_connLink = null;
		m_channel = null;
	}
```

Reply:
sent msg to discuss further